### ISSUE TITLE
`state_dict` and `load_state_dict` for `DataLoader` and `Sampler` to continue training at specific epoch and batch

### ISSUE THREAD

**Issue by visionscaper** [pytorch/pytorch#36650]
## 🚀 Feature
Based on training checkpoint data, allow training to restart at the epoch and **precise** batch, after the last batch trained on when the training checkpoint data was saved.

`state_dict` and `load_state_dict`

## Motivation

I'm training (relatively) large sequence to sequence models (> 100M parameters) with a (relatively large) dataset (1M conversations), using limited training hardware (one machine with 4x 1080Ti); one epoch can easily take 12 hours or (much) longer.

There are situations where I want to restart training based on a training checkpoint I saved during the training process that was stopped before, for some reason. In order to not lose many hours of training time, and to continue in exactly the same way, as if I would if the training process had not stopped, I want to be able to continue at the epoch and **precise** batch, after the last batch trained on when the training checkpoint data was saved.

## Pitch

All relevant components in of a PyTorch training process, such as `nn.Module`s, optimizers and schedulers have `state_dict` and `load_state_dict` methods in order to retrieve and load its states, such that training can be restarted with the exactly the same state as it was stopped.

The only exception, as far as I can tell, is for `DataLoader` and `Sampler`. These (base) classes do not have `state_dict` and `load_state_dict` methods.

It would be very consistent and concise, if a PyTorch user could do the following in order to save and load training state:

1. Get the state of your `DataLoader`s, models, optimizers and schedulers and save it together with any other training process variables (e.g. current epoch and batch index) as a training checkpoint.

2. Load your training checkpoint and subsequently load the state of `DataLoader`s, models, optimizers and schedulers

3. Restart your training loop at the loaded epoch index and batch index and continue training

Please note that it is assumed that the `Sampler` states is nested as a part of the `DataLoader` states.

## Alternatives

`DistributedSampler` has the possibility to set the current epoch, such that the random shuffling of samples is deterministic. However, I can't tell the DataLoader from which specific batch to continue.

**Comment by visionscaper** [comment]
^ Added an `Additional thoughts` section in the feature request description.

**Comment by mruberry** [comment]
Hey Visionscaper! Cool idea. It'd be great to get a PR for this!

**Comment by spthm** [comment]
If this isn't already being worked on, I'd like to have a go at adding `state_dict` and `load_state_dict` to the samplers.

@mruberry would a PR implementing `[load_]state_dict` for only the samplers be welcome/useful? A following PR could implement additional state for `DataLoader`, including serializing their sampler instances - I'd be happy to do that, too, but would be more comfortable starting with only the relatively-simpler `Sampler` classes.

**Comment by mruberry** [comment]
Hey @spthm, this is a better question for @VitalyFedyunin, who's investigating dataloading and dataloaders now

**Comment by VitalyFedyunin** [comment]
We are planned this feature for H2 2021. RFC will be ready in about one moth. Feature shipped approx end of this year.

**Comment by visionscaper** [comment]
This is great news @VitalyFedyunin, thanks for your efforts! 🎉

**Comment by Revist** [comment]
What is the status of this issue? @VitalyFedyunin

**Comment by Robotatron** [comment]
Is it done?

**Comment by visionscaper** [comment]
@VitalyFedyunin I can imagine. I’m wondering what your approach is; it also depends on the type of dataset you provide to the loader, maybe? In any case, really awesome that you are working on this feature, thanks!

**Comment by chuston** [comment]
+1 Vote for this. Suffering through very occasional random “SegmentationFault”* in dataloader workers, training might run for 5 hours before blowing up or it might complete (total run times are several days). Need to restart exactly at last checkpoint. Checkpoints currently save python.random, numpy.random and torch.random random generator states. Can save the shuffled indexes in the initial checkpoint with `index_order = list(the_dataloader.sampler)`.  Would be nice to avoid wholesale replacement of Dataloader just to restart from where we left off.

Thanks for all your hard work.

* thinking segfault from open cv but having a hard time tracking it down

**Comment by Pierre-Bartet** [comment]
I just had to do it and found out that in fact it is the iterator over the dataloader that needs to have a `state_dict` instead of the dataloader itself:

``` python
dataset = MyDataset(10)

dataset_iterator = iter(dataset)

checkpoints = []

for sample in dataset_iterator:
    state = dataset_iterator.state_dict()
    checkpoints.append(state)

dataset_iterator.load_state_dict(checkpoints[5])
list(dataset_iterator)  # -> [6, 7, 8, 9]
```

Here is the modification to `Dataset` that allows it:

``` python
from torch.utils.data import Dataset as _Dataset


class Dataset(_Dataset):
    def __iter__(self):
        return DatasetIterator(self)


class DatasetIterator:
    def __init__(self, dataset, index=0):
        self._dataset = dataset
        self._index = index

    def state_dict(self):
        return {"index": self._index}

    def load_state_dict(self, state):
        self._index = state["index"]

    def __iter__(self):
        return self

    def __next__(self):
        try:
            sample = self._dataset[self._index]
        except IndexError:
            raise StopIteration()

        self._index += 1

        return sample
    

class MyDataset(Dataset):
    def __init__(self, n):
        self._data = list(range(n))

    def __len__(self):
        return len(self._data)

    def __getitem__(self, item):
        return self._data[item]
```

Doing the same thing for `Dataloader` looks extremely complicated as it is a super complex object that does (too?) many things (and as a matter of fact this issue has been opened for years).
I often end up ditching (almost) completely dataloaders to use only datasets:

```python
dataset = MyDataset(10)

dataset = Shuffle(dataset, random_state=0)
dataset = Batches(dataset, batch_size=3, drop_last=True)
dataset = Map(dataset, default_collate)

dataset_iterator = iter(dataset)

dataset_iterator.load_state_dict(some_checkpoint)

for batch in dataset_iterator:
    print(batch)
```

So that `Dataloader` is used "only" for multiprocessing.

**Comment by matthijsvk** [comment]
@VitalyFedyunin any updates?
Do you know if the rewrite of data pipelines (will) support this?

**Comment by Pierre-Bartet** [comment]
According to the TorchData [documentation](https://pytorch.org/data/main/dataloader2.html)  it is some work in progress:
> Snapshot the state of data-preprocessing pipeline (WIP)

**Comment by jramapuram** [comment]
Isn't torchdata depricated now?

**Comment by jiawen-steven-liu** [comment]
@VitalyFedyunin any updates on this feature request and effort?

**Comment by hiyyg** [comment]
https://github.com/pytorch/data/tree/main/torchdata/stateful_dataloader

**Comment by Pierre-Bartet** [comment]
Awesome! Is there any reason not to implement this directly into pytorch?

**Comment by hiyyg** [comment]
I guess it is experimenting. Maybe it will be moved into pytorch after everything is well tested.
